kms/src/main/java/io/kroxylicious/kms/service/Kms.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.kms.service;
+
+import java.util.concurrent.CompletionStage;
+
+import javax.crypto.SecretKey;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+/**
+ * Abstracts the KMS operations needed for Envelope Encryption
+ * @param <K> The type of Key Encryption Key id.
+ * @param <E> The type of encrypted Data Encryption Key.
+ */
+public interface Kms<K, E> {
+
+    /**
+     * Asynchronously generates a new Data Encryption Key, returning it encrypted with the Key Encryption Key given by {@code kekRef}.
+     * The returned DEK can later be decrypted with {@link Kms#decryptEdek(Object, Object)}.
+     * @param kekRef The key encryption key used to encrypt the generated data encryption key.
+     * @return A completion stage for the wrapped data encryption key.
+     * @throws UnknownKeyException If the kek was not known to this KMS.
+     * @throws InvalidKeyUsageException If the given kek was not intended for key wrapping.
+     * @throws KmsException For other exceptions.
+     */
+    @NonNull
+    CompletionStage<E> generateDek(@NonNull K kekRef);
+
+    /**
+     * Asynchronously generates a new Data Encryption Key (DEK) and returns it together with the same DEK wrapped by the Key Encryption Key (KEK) given
+     * by the {@code kekRef},
+     * The returned encrypted DEK can later be decrypted with {@link Kms#decryptEdek(Object, Object)}.
+     * @param kekRef The key encryption key used to encrypt the generated data encryption key.
+     * @return A completion stage for the wrapped data encryption key.
+     * @throws UnknownKeyException If the kek was not known to this KMS.
+     * @throws InvalidKeyUsageException If the given kek was not intended for key wrapping.
+     * @throws KmsException For other exceptions.
+     */
+    @NonNull
+    CompletionStage<DekPair<E>> generateDekPair(@NonNull K kekRef);
+
+    /**
+     * Asynchronously decrypts a data encryption key that was {@linkplain #generateDek(Object) previously encrypted}.
+     * @param kek The key encryption key.
+     * @param edek The encrypted data encryption key.
+     * @return A completion stage for the data encryption key
+     * @throws UnknownKeyException If the kek was not known to this KMS.
+     * @throws InvalidKeyUsageException If the given kek was not intended for key wrapping.
+     * @throws KmsException For other exceptions
+     */
+    @NonNull
+    CompletionStage<SecretKey> decryptEdek(@NonNull K kek, @NonNull E edek);
```

Comment:
Trivial point: This should be `kekId`, not `kek`.

Deeper point: Should this really take/require the `kekId`? By accepting the `kekId` here we're bringing certain expectations about kekId, namely that they _immutably_ refer to the key material needed to decrypt `edek`. That semantic is not compatible with key aliases, which are mutable by intent. In other words, if we stuck with this API we ought to define what happens if I use a key alias as a `kekId` in `generateDek()` (or `generateDekPair()`), change the key that the alias is pointing at, and then use that same `kekId` (now referring to different key material) in this method. The best we can do is fail in some specified way. 

The alternative is the AWS model, where the immutable `kekId` is implicitly serialized within the `edek`, so only the `edek` is needed in this method signature. The provides a strong hint the implementor that immutable key identification is their problem, and that they need to resolve mutable key identifiers to immutable ones when doing `generateDek()`. That then avoids this method failing for this reason, and makes the API more workable with mutable key identifiers like aliases.

I'm strongly inclined to make this change, any objectors?

Reply:
> The alternative is the AWS model, where the immutable kekId is implicitly serialized within the edek.
+100

It makes the responsibilities more clear and gives useful freedom to KMS implementations to store exactly what they need.  Using AWS as an example, its cipertext blob returned by [GenerateDataKey](https://docs.aws.amazon.com/kms/latest/APIReference/API_GenerateDataKey.html) [actually encodes the keyid/version itself] (https://docs.aws.amazon.com/kms/latest/APIReference/API_Decrypt.html#API_Decrypt_RequestSyntax:~:text=KMS%20key%20from%20metadata%20that%20it%20adds%20to%20the%20symmetric%20ciphertext%20blob) used to generate it, so you _could_ make an argument for our AWS implementation not to serialise the KEK itself.   That would save bytes.

Whether we'd actually choose to do that our not is a conversation for a another day (AWS makes best-practice argument against it).